odoo/addons/base/tests/test_ir_cron.py

```diff
@@ -205,61 +192,313 @@ def test_cron_keep_future_triggers(self):
+            (        nothing,             0,   False,          1,          0,          0,  True),
+            (        nothing,             4,   False,          1,          0,          0,  True),
+            ( eleven_success,             0,    True,         10,         10,          0,  True),
+            ( eleven_success,             4,    True,         10,         10,          0,  True),
+            (   five_success,             0,   False,          5,          5,          0,  True),
+            (   five_success,             4,   False,          5,          5,          0,  True),
+            (        failure,             0,   False,          1,          0,          1,  True),
+            (        failure,             4,   False,          1,          0,          0, False),
+            (failure_partial,             0,   False,          5,          5,          1,  True),
+            (failure_partial,             4,   False,          5,          5,          0, False),
+            (  failure_fully,             0,   False,          1,          1,          1,  True),
+            (  failure_fully,             4,   False,          1,          1,          0, False),
         ]
 
-        for setup, expect in matrix:
-            with self.subTest(setup=setup, expect=expect):
+        for cb, curr_failures, trigger, call_count, done_count, fail_count, active in CASES:
+            with self.subTest(cb=cb, failure=curr_failures), closing(self.cr.savepoint()):
                 self.cron.write({
                     'active': True,
-                    'doall': setup.doall,
-                    'numbercall': setup.numbercall,
-                    'nextcall': fields.Datetime.now() - timedelta(days=setup.missedcall - 1),
+                    'failure_count': curr_failures,
+                    'first_failure_date': ten_days_ago if curr_failures else None
                 })
                 with self.capture_triggers(self.cron.id) as capture:
-                    if setup.trigger:
+                    if trigger:
                         self.cron._trigger()
 
-                self.cron.flush_recordset()
-                capture.records.flush_recordset()
+                self.env.flush_all()
                 self.registry.enter_test_mode(self.cr)
+                cb, state = cb(self.cron)
                 try:
-                    with patch.object(self.registry['ir.cron'], '_callback') as callback:
+                    with mute_logger('odoo.addons.base.models.ir_cron'),\
+                            patch.object(self.registry['ir.actions.server'], 'run', cb):
                         self.registry['ir.cron']._process_job(
                             self.registry.db_name,
                             self.registry.cursor(),
-                            self.cron.read(load=None)[0]
+                            {**self.cron.read(load=None)[0], **default_progress_values}
```

Comment:
Dicts support the union `|` operator since [3.9](http://docs.localhost/python/whatsnew/3.9.html#dictionary-merge-update-operators)

![image](https://github.com/odoo/odoo/assets/8208953/1fc75c44-e4f3-48d3-8936-c476cc98f11f)

```suggestion
                            self.cron.read(load=None)[0] | default_progress_values
```

But I see this snippet quite a lot. I would suggest we make a new helper method (that takes an extra `**kwarks` for `progress_id=progress.id`)